examples/thermostat/thermostat-common/include/atomic-write-manager.h

```diff
@@ -0,0 +1,186 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/clusters/thermostat-server/atomic-write.h>
+#include <app/clusters/thermostat-server/thermostat-delegate.h>
+
+#include "thermostat-delegate-impl.h"
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Thermostat {
+
+enum AtomicWriteState
+{
+    kAtomicWriteState_Closed = 0,
+    kAtomicWriteState_Open,
+};
+
+class ThermostatAtomicWriteManager : public AtomicWriteManager
+{
+public:
+    static inline ThermostatAtomicWriteManager & GetInstance() { return sInstance; }
+
+    /**
+     * @brief Set the Delegate object
+     *
+     * @param delegate
+     */
+    void SetDelegate(AtomicWriteDelegate * delegate) override { mDelegate = delegate; };
+
+    /**
+     * @brief Check if there is an open atomic write on an endpoint, optionally associated with a specific attribute
+     *
+     * @param attributeId Optional attribute filter
+     * @param endpoint The endpoint to check atomic write state
+     * @return true if there is an open atomic write
+     * @return false if there is not at open atomic write
+     */
+    bool InWrite(const std::optional<AttributeId> attributeId, EndpointId endpoint) override;
+
+    /**
+     * @brief Check if there is an open atomic write on an endpoint, associated with a given SubjectDescriptor, and optionally
+     * associated with a specific attribute
+     *
+     * @param attributeId Optional attribute filter
+     * @param subjectDescriptor The subject descriptor to check against
+     * @param endpoint The endpoint to check atomic write state
+     * @return true if there is an open atomic write
+     * @return false if there is not an open atomic write
+     */
+    bool InWrite(const std::optional<AttributeId> attributeId, const Access::SubjectDescriptor & subjectDescriptor,
+                 EndpointId endpoint) override;
+
+    /**
+     * @brief Check if there is an open atomic write on an endpoint, associated with the source node for a given command invocation,
+     * and optionally associated with a specific attribute
+     *
+     * @param attributeId
+     * @param commandObj
+     * @param endpoint
+     * @return true
+     * @return false
+     */
+    bool InWrite(const std::optional<AttributeId> attributeId, CommandHandler * commandObj, EndpointId endpoint) override;
```

Comment:
can we have the scoped node id as an argument instead of the commandhandler? i assume we do not need the entire commandhandler object? also how is this different from the subject descriptor API? both boil down to a scoped node id IMO. is that not correct?

Reply:
They both end up comparing a scoped node id, but they fetch that node ID in different ways. If this just took a scoped node ID, then either I'd have to expose `ScopedNodeId GetSourceScopedNodeId(CommandHandler * commandObj)`, in which case I still have an API that takes a CommandHandler, or I need to trust that anyone calling this API will get the scoped node ID in the same fashion.

Reply:
Yeah I agree that you get scoped node id from the two objects in different ways that doesn't justify passing the whole commandObj object or the subjectDescriptor objects if all you need is the scoped node id. Also that reduces > 1 apis to 1 consolidated API. I think it would be a good idea to do so